include/nlohmann/detail/meta/type_traits.hpp

```diff
@@ -68,6 +68,52 @@ using from_json_function = decltype(T::from_json(std::declval<Args>()...));
 template <typename T, typename U>
 using get_template_function = decltype(std::declval<T>().template get<U>());
 
+// trait checking if JSONSerializer<T>::from_json(json const&, udt&) exists
+template <typename BasicJsonType, typename T, typename = void>
+struct has_from_json : std::false_type {};
+
+template <typename BasicJsonType, typename T>
+struct has_from_json<BasicJsonType, T,
+           enable_if_t<not is_basic_json<T>::value>>
+{
+    using serializer = typename BasicJsonType::template json_serializer<T, void>;
+
+    static constexpr bool value =
+        is_detected_exact<void, from_json_function, serializer,
+        const BasicJsonType&, T&>::value;
+};
+
+// This trait checks if JSONSerializer<T>::from_json(json const&) exists
+// this overload is used for non-default-constructible user-defined-types
+template <typename BasicJsonType, typename T, typename = void>
```

Comment:
Maybe add a comment on whether or not 'T' is expected to be a concrete type (string) or is allowed to be a CV-qualified type (string const&). It's not clear if these need std::decay's in there or whether the layer above it should add them?

Reply:
IIRC, the only places where CV-qualified types are specified in the current code is in `is_detected` related code.

I can add a comment at the top of the file.